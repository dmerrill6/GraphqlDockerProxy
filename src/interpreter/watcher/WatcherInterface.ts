
// @flow
import { Endpoints } from '../endpoints';
import * as cloner from 'cloner';

type dataUpdatedListener = (data: Endpoints) => void;

class WatcherInterface{
  endpoints: Endpoints = {};

  constructor() {}

  dataUpdatedListener: dataUpdatedListener;

  watchEndpoint = () => {
    winston.error('you have to override watchEndpoint');
  }

  setDataUpdatedListener = (listener:dataUpdatedListener) => {
    this.dataUpdatedListener = listener;
  }

  callDataUpdateListener = async() => {
    const realEndpoint = cloner.deep.copy(this.endpoints);
    for (const one in realEndpoint) {
      if (realEndpoint[one].length === 0) {
        delete realEndpoint[one];
      }

    }
    this.dataUpdatedListener(realEndpoint);
  }

  deleteEndpoint = (namespace: string, uniqueIdentifier: string) => {
    if (this.endpoints[namespace] === undefined) {
      return;
    }
    for (let i = 0 ; i < this.endpoints[namespace].length; i = i + 1) {
      if (this.endpoints[namespace][i].__deploymentName === uniqueIdentifier) {
        this.endpoints[namespace].splice(i, 1);
      }
    }
    if (this.endpoints[namespace].length === 0) {
      delete this.endpoints[namespace];
    }
  }
}

export { WatcherInterface };
