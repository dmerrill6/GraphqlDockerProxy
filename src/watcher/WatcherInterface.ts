
// @flow
import { Endpoints } from '../finder/findEndpointsInterface';

type dataUpdatedListener = (data: Endpoints) => void;

class WatcherInterface{
  constructor() {}

  dataUpdatedListener: dataUpdatedListener;

  watchEndpoint = () => {
    winston.error('you have to override watchEndpoint');
  }

  setDataUpdatedListener = (listener:dataUpdatedListener) => {
    this.dataUpdatedListener = listener;
  }
}

export { WatcherInterface };
