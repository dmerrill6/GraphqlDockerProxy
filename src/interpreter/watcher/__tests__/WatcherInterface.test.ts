import { WatcherInterface } from '../WatcherInterface'

describe('Test the WatcherInterface', () => {
  let watcher : WatcherInterface = null
  beforeEach(() => {
    watcher = new WatcherInterface()
  })
  it('snapshot all methodes and const', () => {
    for (const one in watcher) {
      expect(one).toMatchSnapshot()
    }
  })

  it('test call watch and abort Stream', () => {
    watcher.watchEndpoint = jest.fn()
    watcher.abortAllStreams = jest.fn()
    watcher.resetConnection()

    expect(watcher.watchEndpoint).toBeCalled()
    expect(watcher.abortAllStreams).toBeCalled()
  })

  it('test deleteEndpoint do the right job', () => {
    const wa = {
      a: [
        {
          url: 'mock',
          namespace:'mock',
          typePrefix: '_mock',
          __deploymentName: 'mock',
          __imageID: 'mock',

        },
        {
          url: 'mock',
          namespace:'mock',
          typePrefix: '_mock',
          __deploymentName: 'delete',
          __imageID: 'mock',
        },
      ],
    }
    watcher.endpoints = wa
    watcher.deleteEndpoint('a', 'delete')
    expect(watcher.endpoints.a.length).toBe(1)
  })

  it('test callDataUpdateListener', async() => {
    const wa = {
      a: [
        {
          url: 'mock',
          namespace:'mock',
          typePrefix: '_mock',
          __deploymentName: 'mock',
          __imageID: 'mock',

        },
      ],
    }
    watcher.endpoints = wa
    const listener = jest.fn()
    watcher.setDataUpdatedListener(listener)
    await watcher.callDataUpdateListener()
    expect(listener).toBeCalledWith(wa)
  })
})
