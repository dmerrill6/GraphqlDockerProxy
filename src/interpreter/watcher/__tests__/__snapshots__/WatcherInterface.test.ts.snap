// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test the WatcherInterface snapshot all methodes and const 1`] = `"resetConnection"`;

exports[`Test the WatcherInterface snapshot all methodes and const 2`] = `"endpoints"`;

exports[`Test the WatcherInterface snapshot all methodes and const 3`] = `"handleRestart"`;

exports[`Test the WatcherInterface snapshot all methodes and const 4`] = `"watchEndpoint"`;

exports[`Test the WatcherInterface snapshot all methodes and const 5`] = `"abortAllStreams"`;

exports[`Test the WatcherInterface snapshot all methodes and const 6`] = `"setDataUpdatedListener"`;

exports[`Test the WatcherInterface snapshot all methodes and const 7`] = `"callDataUpdateListener"`;

exports[`Test the WatcherInterface snapshot all methodes and const 8`] = `"deleteEndpoint"`;

exports[`Test the WatcherInterface snapshot all methodes and const 9`] = `"constructor"`;
